language/go/generate.go

```diff
@@ -427,53 +478,125 @@ func (g *generator) generateProto(mode proto.Mode, target protoTarget, importPat
 
 	gc := getGoConfig(g.c)
 	filegroupName := legacyProtoFilegroupName
-	protoName := target.name
+	var protoName string
+	if len(targets) == 1 {
+		protoName = targets[0].name
+	}
 	if protoName == "" {
-		importPath := InferImportPath(g.c, g.rel)
+		if importPath == "" {
+			importPath = InferImportPath(g.c, g.rel)
+		}
 		protoName = proto.RuleName(importPath)
 	}
 	goProtoName := strings.TrimSuffix(protoName, "_proto") + goProtoSuffix
 	visibility := g.commonVisibility(importPath)
 
 	if mode == proto.LegacyMode {
 		filegroup := rule.NewRule("filegroup", filegroupName)
-		if target.sources.isEmpty() {
+		if targets[0].sources.isEmpty() {
 			return "", []*rule.Rule{filegroup}
 		}
-		filegroup.SetAttr("srcs", target.sources.build())
+		filegroup.SetAttr("srcs", targets[0].sources.build())
 		if g.shouldSetVisibility {
 			filegroup.SetAttr("visibility", visibility)
 		}
 		return "", []*rule.Rule{filegroup}
 	}
 
-	if target.sources.isEmpty() {
+	var atLeastOneTargetHasSources bool
+	for _, target := range targets {
+		if !target.sources.isEmpty() {
+			atLeastOneTargetHasSources = true
+			break
+		}
+	}
+
+	if !atLeastOneTargetHasSources {
 		return "", []*rule.Rule{
 			rule.NewRule("filegroup", filegroupName),
 			rule.NewRule("go_proto_library", goProtoName),
 		}
 	}
 
 	goProtoLibrary := rule.NewRule("go_proto_library", goProtoName)
-	goProtoLibrary.SetAttr("proto", ":"+protoName)
+	if len(targets) == 1 {
+		goProtoLibrary.SetAttr("proto", ":"+protoName)
+	} else {
+		// generate protoNames with prefix ":"
+		protoNames := make([]string, len(targets))
+		for i := range targets {
+			protoNames[i] = ":" + targets[i].name
+		}
+		goProtoLibrary.SetAttr("protos", protoNames)
+	}
+
 	g.setImportAttrs(goProtoLibrary, importPath)
-	if target.hasServices {
+	var atLeastOneTargetHasServices bool
+	for _, target := range targets {
+		if target.hasServices {
+			atLeastOneTargetHasServices = true
+			break
+		}
+	}
+	if atLeastOneTargetHasServices {
 		goProtoLibrary.SetAttr("compilers", gc.goGrpcCompilers)
 	} else if gc.goProtoCompilersSet {
 		goProtoLibrary.SetAttr("compilers", gc.goProtoCompilers)
 	}
 	if g.shouldSetVisibility {
 		goProtoLibrary.SetAttr("visibility", visibility)
 	}
-	goProtoLibrary.SetPrivateAttr(config.GazelleImportsKey, target.imports.build())
+	if len(targets) == 1 {
+		goProtoLibrary.SetPrivateAttr(config.GazelleImportsKey, targets[0].imports.build())
+	} else {
+		protoSources := make(map[string]bool)
```

Comment:
nit: prefer `map[string]struct{}`, because each entry in map[string]bool has 3 states: not exist, true, false, but we only need to know whether it exists or not.

Reply:
I don't think I understand. `protoSources` is a set being used to dedupe proto src files, which I implemented as a map. Why would there need to be 3 states then?